utils/env_preferences/src/apple.rs

```diff
@@ -2,135 +2,203 @@
-        // https://developer.apple.com/documentation/corefoundation/cfarrayref. It is ensured that `locale_carr_ref` is not mutated
-        // Immutablility ensures that nothing is overriden during it's scope
-        let locale_carr_ref = unsafe { CFLocaleCopyPreferredLanguages() };
-
-        if !locale_carr_ref.is_null() {
-            // SAFETY: The call to `CFArrayGetCount` is only made when is `locale_carr_ref` is not `NULL`
-            let count = unsafe { CFArrayGetCount(locale_carr_ref as _) };
-            for i in 0..count {
-                // SAFETY: The call to `CFArrayGetValueAtIndex` is safe because we are iterating from 0 to count (`CFArrayGetCount`) which
-                // in itself will always be greater than 0 and less than count ensuring we will not get "out of bounds" error
-                let lang_ptr = unsafe { CFArrayGetValueAtIndex(locale_carr_ref as _, i) };
-
-                if !lang_ptr.is_null() {
-                    // SAFETY: The call to `CFStringGetCStringPtr` because the reference of string we are accessing is not `NULL`
-                    // Returns pointer in O(1) without any memory allocation. This can return NULL so to handle it
-                    let lang_str = unsafe { CFStringGetCStringPtr(lang_ptr as _, 0) };
-
-                    if !lang_str.is_null() {
-                        // SAFETY: A valid `NULL` terminator is present which is a requirement of `from_ptr`
-                        let lang_rust_str = unsafe { CStr::from_ptr(lang_str) }
-                            .to_str()
-                            .unwrap_or("Unknown")
-                            .to_string();
-                        languages.push(lang_rust_str);
-                    }
-
-                    // TODO: In case `lang_str` is `NULL` try retrieving the string using `CFStringGetCString`
-                    // Ref: https://developer.apple.com/documentation/corefoundation/1542721-cfstringgetcstring?language=objc
-                    // Note: Not optimal and may give inconsistent results if buffer is not large enough, must add sanity checks
-                    // whenever implemented
+use core_foundation_sys::{
+    array::{CFArrayGetCount, CFArrayGetValueAtIndex},
+    base::{CFIndex, CFRelease, CFRetain},
+    calendar::{CFCalendarCopyCurrent, CFCalendarCopyLocale, CFCalendarGetIdentifier},
+    locale::{CFLocaleCopyPreferredLanguages, CFLocaleGetIdentifier},
+    string::{
+        kCFStringEncodingUTF8, CFStringGetCString, CFStringGetCStringPtr, CFStringGetLength,
+        CFStringRef,
+    },
+    timezone,
+};
+use libc::c_char;
+use std::ffi::CStr;
+
+use crate::RetrievalError;
+
+// Helper function
+fn get_string(ptr: CFStringRef) -> String {
+    // SAFETY: It returns length of the string, from above conditional statement we ensure
+    // that the `lang_ptr` is not NULL thus making it safe to call
+    let length = unsafe { CFStringGetLength(ptr) as usize };
+
+    let mut c_str_buf = vec![0; length * 4];
```

Comment:
issue: why is this `length * 4`? I think this should be `length + 1`: https://developer.apple.com/documentation/corefoundation/1542721-cfstringgetcstring#parameters

Reply:
*4 is an overkill 😅

Reply:
it's underkill and maybe UB if the string is empty.